### PR TITLE
Add perimeter to morphometrics 

### LINKED ISSUE #497: Add axon and axonmyelin perimeters to morphometrics
**Context**
In addition to the "circle area" and "ellipse minor axis" methods to compute axon diameter, another method found in the litterature ([Bartmeyer et al, 2021](https://www.nature.com/articles/s41598-021-82575-9)) and used by one of our collaborator is to use the inner and outer myelin perimeter.
With this method, the diameter is computed as the diameter of a circle with the same perimeter as the region.

**Implementation**
As a first step, we want to add the axon perimeter and axonmyelin perimeter to the morphometrics excel file.
The perimeter can be extracted from skimage regionprops.

**Future steps**
We could include this third method to compute morphometrics in addition to "circle area" and "ellipse minor axis".

### PR DESCRIPTION
**Pull request by vs74**
## Checklist

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've added relevant tests for my contribution
- [ ] I've updated the documentation and/or added correct docstrings
- [ ] I've assigned a reviewer
- [x] I've consulted [ADS's internal developer documentation](https://github.com/neuropoly/axondeepseg/wiki/Guidelines-for-contribution) to ensure my contribution is in line with any relevant design decisions

## Description
This PR aims to add `axon `and `axonmyelin` perimeter in the morphometrics file. 

## Linked issues
Fixes #497

### CODE REVIEW

File: ads_plugin.py [hunk 1]
```diff
@@ -601,7 +603,7 @@ def on_compute_morphometrics_button(self, event):
 
         # Compute statistics
         stats_array = get_axon_morphometrics(im_axon=pred_axon, im_myelin=pred_myelin, pixel_size=pixel_size)
-
+        # print(stats_array)
```
Comment (mathieuboudreau) [inline ads_plugin.py:606]: Is this intentional to have the print here, or was it used for debug and should be removed?

File: AxonDeepSeg/morphometrics/compute_morphometrics.py [hunk 2]
```diff
@@ -110,13 +110,16 @@ def get_axon_morphometrics(im_axon, path_folder=None, im_myelin=None, pixel_size
         axon_diam = prop_axon.equivalent_diameter * pixelsize
         # Axon area in µm^2
         axon_area = prop_axon.area * (pixelsize ** 2)
+        # Axon perimeter in micrometers
+        axon_perimeter = prop_axon.perimeter * pixelsize
```
Comment (jcohenadad) [inline AxonDeepSeg/morphometrics/compute_morphometrics.py:114]: what if anisotropic pixel?

Comment (mathieuboudreau) [inline AxonDeepSeg/morphometrics/compute_morphometrics.py:114]: In terms of the myelin, is this the only the inner perimeter? If so, it should be documented here for clarity I think.

Comment (vs74) [reply AxonDeepSeg/morphometrics/compute_morphometrics.py:114]: Fixed

File: ads_plugin.py (hunk 1 above)
Comment (vs74) [reply ads_plugin.py]: It was used for debugging, thanks for catching it. I fixed it.

**Comment by vs74** [comment]
I added tests for validating this feature on the simulated axon-myelin instances. Here is the simulated image where axon shape is circular on which I tested this PR. 


image


The inner perimeter ( **axon perimeter**) and the outer perimeter **(axon myelin perimeter**) of the simulated axon are equivalent to the perimeters computed by the skimage regionprops inbuilt perimeter attribute. 


However, note the functionality was only tested for axon shape being **circular** in nature.

**Comment by vs74** [comment]
**TEST 2
________**

For ellipsoidal shape, since there is no way to compute the morphometrics, I manually computed the perimeter.

Here is the simulated image. 
image





**The ground truth perimeter computed manually using ImageJ on this simulated image**

Instance   no | Axon Perimeter | Axonmyelin Perimeter
-- | -- | --
1 | 317.301 | 351.858
2 | 289.035 | 361.31
3 | 263.978 | 377.05
4 | 245.296 | 402.621
5 | 225.616 | 441.851
6 | 190.714 | 473.511
7 | 180.503 | 586.685
8 | 157.717 | 788.764



**Perimeter computed using this PR (**skimage regionprops inbuilt perimeter attribute**)**



Instance no | Axon Perimeter | Axonmyelin Perimeter
-- | -- | --
1 | 326.391919 | 365.019336
2 | 297.4213562 | 374.67619
3 | 270.1076477 | 388.333044
4 | 246.7939392 | 415.646753
5 | 227.4802307 | 456.617316
6 | 192.8528137 | 490.27417
7 | 179.882251 | 608.901587
8 | 162.9116882 | 821.185858

**Comment by mathieuboudreau** [comment]
Looks close enough for me for the ellipse case. You could do a % difference comparison to give us an order of magnitude of the possible error in terms of manual vs. our tool, my guess is it will be around 2-3% which I think is sufficiently low.

**Comment by vs74** [comment]
You are right @mathieuboudreau 



Instance   no | Axon Perimeter | Axonmyelin Perimeter
-- | -- | --
0 | 2.86507732 | 3.74052487
1 | 2.90150198 | 3.69936903
2 | 2.32202977 | 2.99245312
3 | 0.61066599 | 3.23523934
4 | 0.862685 | 3.34192199
5 | 1.12147705 | 3.54018597
6 | 0.3450863 | 3.7867999
7 | 3.2936768 | 4.11046374

**Review by mariehbourget — CHANGES_REQUESTED** [review]
Thanks Vasudev for implementing this!
I tested it, and it works as expected  in FSLeyes. 
However, the axon_perimeter and axonmyelin_perimeter are missing from the stats array in CLI [here](https://github.com/neuropoly/axondeepseg/blob/f0a00067a5aa4477c5a05eabe796a2e5f212145d/AxonDeepSeg/morphometrics/launch_morphometrics_computation.py#L144) and  [here](https://github.com/neuropoly/axondeepseg/blob/f0a00067a5aa4477c5a05eabe796a2e5f212145d/AxonDeepSeg/morphometrics/launch_morphometrics_computation.py#L164).


I compared a few perimeter measurements between the morpho files and measures in ImageJ on the SEM data test. All the measurements I checked where within 1 microns between ADS and ImageJ.

**Comment by vs74** [comment]
> However, the axon_perimeter and axonmyelin_perimeter are missing from the stats array in CLI [here](https://github.com/neuropoly/axondeepseg/blob/f0a00067a5aa4477c5a05eabe796a2e5f212145d/AxonDeepSeg/morphometrics/launch_morphometrics_computation.py#L144) and [here](https://github.com/neuropoly/axondeepseg/blob/f0a00067a5aa4477c5a05eabe796a2e5f212145d/AxonDeepSeg/morphometrics/launch_morphometrics_computation.py#L164).


I actually forgot that we had merged the morphometrics CLI into ADS, so skipped adding the functionality there. Thanks @mariehbourget for catching it. I've fixed this and you can now test `axondeepseg_morphometrics` CLI.

File: AxonDeepSeg/morphometrics/launch_morphometrics_computation.py [hunk 3]
```diff
@@ -146,7 +146,9 @@ def main(argv=None):
                                         ('y0', 'f4'),
                                         ('gratio', 'f4'),
                                         ('axon_area', 'f4'),
+                                        ('axon_perimeter', 'f4'),
                                         ('myelin_area', 'f4'),
+                                        ('axonmyelin_perimeter', 'f4'),
```
Comment (mariehbourget) [inline AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:151]: We should have the stats in the same order as in the fsleyes plugin.

File: AxonDeepSeg/morphometrics/launch_morphometrics_computation.py [hunk 4]
```diff
@@ -167,7 +169,9 @@ def main(argv=None):
                             stats['y0'],
                             stats['gratio'],
                             stats['axon_area'],
+                            stats['axon_perimeter'],
                             stats['myelin_area'],
+                            stats['axonmyelin_perimeter'],
```
Comment (mariehbourget) [inline AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:174]: Same comment as above regarding the order.

**Review by mariehbourget — COMMENTED** [review]
Thanks @vs74, the CLI works as expected.
However, I would order the stats in the same way in CLI and FSLeyes.

File: AxonDeepSeg/morphometrics/launch_morphometrics_computation.py (hunk 3 above)
Comment (vs74) [reply AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:151]: I second on that. Fixed in the latest commit.

**Review by mariehbourget — APPROVED** [review]
I tested again, both files from CLI and fsleyes are identical.
Thanks @vs74!

**Comment by vs74** [comment]
> I tested again, both files from CLI and fsleyes are identical.
> Thanks @vs74!

Thanks a ton, @mariehbourget for thoroughly testing this PR. I'm merging this PR.

**Pull request merged**